src/memory.zig

```diff
@@ -65,48 +65,61 @@ fn deinitIsVoid(comptime T: type) bool {
     };
 }
 
-/// Calls `deinit` on `ptr_or_slice`, or on every element of `ptr_or_slice`, if such a `deinit`
-/// method exists.
+/// Calls `deinit` on `ptr_or_slice`, or on every element of `ptr_or_slice`, if the pointer points
+/// to a struct or tagged union.
 ///
 /// This function first does the following:
 ///
-/// * If `ptr_or_slice` is a single-item pointer, calls `ptr_or_slice.deinit()`, if that method
-///   exists.
-/// * If `ptr_or_slice` is a slice, calls `deinit` on every element of the slice, if the slice
-///   elements have a `deinit` method.
+/// * If `ptr_or_slice` is a single-item pointer of type `*T`:
+///   - If `T` is a struct or tagged union, calls `ptr_or_slice.deinit()`
+///   - If `T` is an optional, checks if `ptr_or_slice` points to a non-null value, and if so,
+///     calls `bun.memory.deinit` with a pointer to the payload.
+/// * If `ptr_or_slice` is a slice, for each element of the slice, calls `bun.memory.deinit` with
+///   a pointer to the element.
 ///
 /// Then, if `ptr_or_slice` is non-const, this function also sets all memory referenced by the
 /// pointer to `undefined`.
 ///
 /// This method does not free `ptr_or_slice` itself.
 pub fn deinit(ptr_or_slice: anytype) void {
     const ptr_info = @typeInfo(@TypeOf(ptr_or_slice));
+    switch (comptime ptr_info.pointer.size) {
+        .slice => {
+            for (ptr_or_slice) |*elem| {
+                deinit(elem);
+            }
+            return;
+        },
+        .one => {},
+        else => @compileError("unsupported pointer type"),
+    }
+
     const Child = ptr_info.pointer.child;
     const mutable = !ptr_info.pointer.is_const;
+    defer {
+        if (comptime mutable) {
+            ptr_or_slice.* = undefined;
+        }
+    }
 
     const needs_deinit = comptime switch (@typeInfo(Child)) {
```

Comment:
we could consider making this switch exhaustive with unhandled types going to a compileError